### DISCUSSION TITLE
How to test services?

### DISCUSSION THREAD

**Discussion by bard** [statelyai/xstate#2204] [Q&A]
The method at https://xstate.js.org/docs/guides/testing.html#testing-services is straightforward when you want to test what happens when 1) the service is in the initial state and 2) you send a single event. 

However, that's a special case. Did anyone manage to apply it when either or neither of those conditions are met?

For example, how would you test what happens when a `{ type: 'CLICK' }` event is received when the state is `error`, or `{ type: 'PAUSE' }` when the state is `playing`, where both the `playing` and `error` states are only reached after a flow involving various events and potentially asynchronous invocations (`setConfig`, `fetchSongInfo`, etc.)?

Things I've tried:

1. Writing a long list of events, a long list of states, and orchestration code that feeds each event to the service and then asserts the state, also taking into account eventless transition. Sort of works, but it's brittle and complex.
2. Start the service in an arbitrary state, send an event, and assert the state, as per documentation. Not working because not supported (related: https://github.com/davidkpiano/xstate/discussions/1683, https://github.com/davidkpiano/xstate/discussions/1515, https://github.com/davidkpiano/xstate/pull/1903). 

Things I've considered:

- Give up testing the service entirely, and only test transitions using `machine.transition`. This seems very granular, and would involve mocking events that are normally generated inside XState, such as `done` and `error` events from promise services, which is reaching inside the system perhaps a bit too much.

**Comment by bard** [comment]
Addendum to things I've tried (or, rather, I'm trying):

- Configure the machine to accept a `{ type: '__BECOME', state: string, context: Context }` in the initial state, as a backdoor to make strategy 2 above work. (I think this was mentioned in one of the issues referenced above, too.)

**Comment by davidkpiano** [comment]
Can you share the code that you're using for this approach (even a reduced example is fine):

> Writing a long list of events, a long list of states, and orchestration code that feeds each event to the service and then asserts the state, also taking into account eventless transition. Sort of works, but it's brittle and complex.

**Reply by bard** [reply]
Here is a simplified version. The machine loads a segmented media resource and plays the segments, interleaved with silence. It's all setup until the line `playerService.onTransition...`:

```ts
import chai, { expect } from 'chai'
import chaiSubset from 'chai-subset'
import { createStudyMachine, Event } from './study'
import { interpret } from 'xstate'
chai.use(chaiSubset)

describe('media machine', () => {
  it('happy path', (done) => {
    const events: (Event | null)[] = [
      {
        type: 'LOAD' as const,
        mediaId: 'media123',
        position: 0,
      },
      {
        type: 'MEDIA_READY',
        segmentCount: 2,
      },
      { type: 'PLAY' },
      { type: 'STEP_FINISHED' },
      null, // AFTER_PLAY delay
      { type: 'STEP_FINISHED' },
      null, // AFTER_PLAY delay
    ]

    const stateFragments = [
      { value: 'empty', context: { media: [], program: null } },
      {
        value: 'loading',
        context: { mediaId: 'media123', position: 0 },
      },
      {
        value: 'idle',
        context: { position: 0, segmentCount: 2 },
      },
      {
        value: { playing: 'media' },
        context: { position: 0 },
      },
      {
        value: { playing: 'silence' },
        context: { position: 0 },
      },
      {
        value: { playing: 'media' },
        context: { position: 1 },
      },
      {
        value: { playing: 'silence' },
        context: { position: 1 },
      },
      {
        value: 'idle',
        context: { position: 0, segmentCount: 2 },
      },
    ]

    const playerService = interpret(
      createStudyMachine().withConfig({
        delays: { AFTER_PLAY: 1 },
        actions: { fetchText: () => {} },
      }),
    ).start()

    playerService.onTransition((state) => {
      const expectedStateFragment = stateFragments.shift()
      if (expectedStateFragment === undefined) {
        const stateDump = JSON.stringify(
          { value: state.value, context: state.context },
          null,
          2,
        )
        return done(
          new Error(
            `Missing state expectation. Current machine state: \n${stateDump}\n`,
          ),
        )
      }

      expect(state.value).to.deep.equal(expectedStateFragment.value)
      expect(state.context).to.containSubset(expectedStateFragment.context)

      const nextEvent = events.shift()
      if (nextEvent) {
        playerService.send(nextEvent)
      } else if (nextEvent === null) {
        // do nothing, expect the machine to transition of its own due to actions
      } else if (nextEvent === undefined) {
        // no events left
        if (stateFragments.length > 0) {
          done(
            new Error(
              `${stateFragments.length} states left to check but no events left to send`,
            ),
          )
        } else {
          done()
        }
      }
    })
  })
})
```

**Reply by bard** [reply]
For completeness, here's also how I'm using the `__BECOME` trick.

Machine definition fragment:

```ts
      on: {
        __BECOME: [
          {
            cond: (_context, event) => event.state === 'loading',
            actions: assign((_context, event) => event.context),
            target: 'loading',
          },
          {
            cond: (_context, event) => event.state === 'idle',
            actions: assign((_context, event) => event.context),
            target: 'idle',
          },
        ],
```

(Needs a `cond` for every state that has to support this. I couldn't see a way to make `target` dynamic.)

Test:

```ts
  it('enters playing state when receiving play event in idle state', (done) => {
    const playerService = interpret(createPlayerMachine())

    playerService
      .onTransition((state) => {
        if (state.matches('playing')) done()
      })
      .start()

    playerService.send({
      type: `__BECOME`,
      state: 'idle',
      context: { mediaId: 'media123', segmentCount: 3 },
    })

    playerService.send({ type: 'PLAY' })
  })
```

After noticing `State.from` mentioned in https://github.com/davidkpiano/xstate/discussions/2213, I decided to give it a go, and it also appears to work, though I couldn't make the typechecker happy with any state other than the initial one:

```ts
  it('enters playing state when receiving play event in idle state (version using State.from)', (done) => {
    const playerService = interpret(createPlayerMachine())

    playerService
      .onTransition((state) => {
        if (state.matches('playing')) done()
      })
      .start(
        State.from('idle', { mediaId: 'media123', segmentCount: 3 }),
      )

    playerService.send({ type: 'PLAY' })
  })
```

**Reply by bard** [reply]
@davidkpiano does any of that look sensible to you?

I was thinking that maybe the reason why the first approach looks complicated is because it really belongs to a library, and a test should import and use it as `expect(service).toSatisfyScenario(states, events)`, sort of how [jest-marbles](https://www.npmjs.com/package/jest-marbles) does for observables.

**Reply by davidkpiano** [reply]
@bard We're currently working on a way to support state restoration for actors. The `.getSnapshot()` method is the first part of that; so ideally, this strategy should work:

> Start the service in an arbitrary state, send an event, and assert the state, as per documentation. Not working because not supported (related: #1683, #1515, #1903).
